### PR TITLE
Add common service for logging deprecated api usage

### PR DESCRIPTION
**Pull request by mjbvz**
For #88391

Adds a new `ExtHostApiDeprecationService`. This service logs a warning and telemetry when a deprecated API is used.

Updates some of the more simple deprecated apis to use this new service

### CODE REVIEW

**Comment by mjbvz** [comment]
This change doesn't cover deprecations for methods such as `TextDocument.show()` or constructors such as `Task`. 

The reason is that I do not know how to get the calling extension information info in the implementation code

File: src/vs/workbench/api/common/extHostApiDeprecationService.ts [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	private readonly _telemetryShape: extHostProtocol.MainThreadTelemetryShape;
+
+	constructor(
+		@IExtHostRpcService rpc: IExtHostRpcService,
+		@ILogService private readonly _extHostLogService: ILogService,
+	) {
+		this._telemetryShape = rpc.getProxy(extHostProtocol.MainContext.MainThreadTelemetry);
+	}
+
+	public report(apiId: string, extension: IExtensionDescription, migrationSuggestion: string): void {
```
Comment (jrieken) [inline src/vs/workbench/api/common/extHostApiDeprecationService.ts]: Use `ExtensionIdentifier` only

File: src/vs/workbench/api/common/extHostApiDeprecationService.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+		type DeprecationTelemetry = {
+			extensionId: string;
+			apiId: string;
+		};
+		type DeprecationTelemetryMeta = {
+			extensionId: { classification: 'SystemMetaData', purpose: 'PerformanceAndHealth' };
+			apiId: { classification: 'SystemMetaData', purpose: 'PerformanceAndHealth' };
+		};
+		this._telemetryShape.$publicLog2<DeprecationTelemetry, DeprecationTelemetryMeta>('resolveCompletionItem/invalid', {
```
Comment (jrieken) [inline src/vs/workbench/api/common/extHostApiDeprecationService.ts]: ⚠️ copy&paste error for the event name

**Review by jrieken — CHANGES_REQUESTED** [review]

**Comment by alexr00** [comment]
For the tasks constructor, I could mark the task as having been created using deprecated api, then in the extHostTaskService when reading tasks use `IExtensionService` to find the extension based on the task definition.

File: src/vs/workbench/api/common/extHostApiDeprecationService.ts (hunk 1 above)
Comment (mjbvz) [reply src/vs/workbench/api/common/extHostApiDeprecationService.ts]: We also want to log a warning message if the extension is in development mode. For that I believe it needs the full `IExtensionDescription`

**Comment by mjbvz** [comment]
@alexr00 I think you may also run a problem getting access to the deprecation service inside the task constructor. If you know some way to workaround this, feel free to give it a try

**Review by roblourens — APPROVED** [review]

**Review by eamodio — APPROVED** [review]

**Pull request merged**